test/unit/math/test_ad.hpp

```diff
@@ -1567,16 +1761,178 @@ void expect_ad_matvar(F&& f, const EigMat& x) {
+  test_matvar_sum_gradient(base_result, mix_result, base_mv1, base_mv2, mix_mv1,
+                           mix_vm2);
+  stan::math::set_zero_all_adjoints();
+  test_matvar_gradient(base_result, mix_result2, base_mv1, base_mv2, mix_vm1,
+                       mix_mv2);
+  stan::math::set_zero_all_adjoints();
+  test_matvar_sum_gradient(base_result, mix_result2, base_mv1, base_mv2,
+                           mix_vm1, mix_mv2);
+}
+/**
+ * For binary function check that an Eigen matrix of vars and a var with an
+ * inner eigen type return the same values and adjoints. This is done by
+ * calling the function on both types, summing the results, then taking the
+ * gradient of the sum which will propogate the adjoints upwards.
+ *
+ * @tparam F A lambda or functor type that calls the unary function.
+ * @tparam Type1 Either arithmetic or `var`. If arithmetic This will make the
+ * LHS of the function invocation be arithmetic, if `var` this will make the LHS
+ * be an Eigen matrix of vars and a var with an inner matrix type.
+ * @tparam Type2 Either arithmetic or `var`. If arithmetic This will make the
+ * RHS of the function invocation be arithmetic, if `var` this will make the RHS
+ * be an Eigen matrix of vars and a var with an inner matrix type.
+ * @tparam EigMat1 An eigen type
+ * @tparam EigMat2 An eigen type
+ * @param f a lambda
+ * @param x An Eigen matrix.
+ * @param y An Eigen matrix.
+ *
+ */
+template <typename Type1, typename Type2, typename F, typename EigMat1,
+          typename EigMat2>
+void expect_ad_matvar_impl(F&& f, const EigMat1& x, const EigMat2& y) {
+  using stan::is_var;
+  using stan::plain_type_t;
+  using stan::math::promote_scalar_t;
+  using stan::math::var;
+  using stan::math::var_value;
+  using stan::math::test::type_name;
+  using mat_var1 = promote_scalar_t<Type1, EigMat1>;
+  using mat_var2 = promote_scalar_t<Type2, EigMat2>;
+  using var_mat1 = std::conditional_t<is_var<Type1>::value,
+                                      var_value<plain_type_t<EigMat1>>,
+                                      plain_type_t<EigMat1>>;
+  using var_mat2 = std::conditional_t<is_var<Type2>::value,
+                                      var_value<plain_type_t<EigMat2>>,
+                                      plain_type_t<EigMat2>>;
+  mat_var1 A_mv1 = x;
+  mat_var2 A_mv2 = y;
+  plain_type_t<decltype(f(A_mv1, A_mv2))> A_mv_f;
+  var_mat1 A_vm1 = x;
+  var_mat2 A_vm2 = y;
+  plain_type_t<decltype(f(A_vm1, A_vm2))> A_vm_f;
+  check_return_type(A_mv_f, A_mv1, A_mv2);
+  check_return_type(A_vm_f, A_vm1, A_vm2);
+  // If one throws, the other should throw as well
+  try {
+    A_mv_f = f(A_mv1, A_mv2);
+  } catch (...) {
+    try {
+      f(A_vm1, A_vm2);
+      FAIL() << type_name<mat_var1>() << " and " << type_name<mat_var2>()
+             << " throws, expected " << type_name<var_mat1>() << " and "
+             << type_name<var_mat2>() << " version to throw";
+    } catch (...) {
+      SUCCEED();
```

Comment:
If both functions have thrown no reason to try to check gradients from here. Just return.